java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java

```diff
@@ -128,6 +132,20 @@ private ProgramState handleMethodInvocation(CheckerContext context, MethodInvoca
     return programState;
   }
 
+  private static ProgramState handleMethodReference(CheckerContext context, MethodReferenceTree mrt) {
+    ProgramState programState = context.getState();
+    if (TERMINAL_OPERATIONS.anyMatch(mrt.method().symbol())) {
+      Tree expression = mrt.expression();
+      if (expression.is(Tree.Kind.IDENTIFIER)) {
+        SymbolicValue ownerSV = programState.getValue(((IdentifierTree) expression).symbol());
+        if (ownerSV != null && programState.getConstraint(ownerSV, StreamPipelineConstraint.class) != null) {
+          return Iterables.getOnlyElement(ownerSV.setConstraint(programState, StreamPipelineConstraint.CONSUMED));
```

Comment:
There should be a check if the SV is already consumed and if it's the case issue should be raised for S3959, see also comment in tests